### DISCUSSION TITLE
import() function

### DISCUSSION THREAD

**Discussion by ghost** [pulsar-edit/.github#185] [Community Packages]
The `import()` function crashes my Pulsar. Is there any way to get it working? It freezes for a second then says "The editor has crashed". I'm on Fedora.

I'm trying to hack out a replacement for [`linter-js-standard`](https://github.com/ricardofbarros/linter-js-standard). It uses an old version of [`standard`](https://github.com/standard/standard). The latest version is pure ESM, so it has to be loaded with an import.

Node.js version is v14.6.0. The docs say [this is supported](https://nodejs.org/docs/v14.6.0/api/esm.html#esm_import_expressions):

>[Dynamic import()](https://wiki.developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Dynamic_Imports) is supported in both CommonJS and ES modules. It can be used to include ES module files from CommonJS code.

**Comment by confused-Techie** [comment]
Writing in `import()` shouldn't crash Pulsar at all. If you are modifying the community package of linter-js-standard` while using it, it's possible that change is causing that package to crash, which while not intended or hoped, community packages can crash Pulsar totally.

Now, if you aren't modifying that package directly, but instead just working in a text file,  there's absolutely no reason that should crash your editor as all, since to Pulsar `import()` is just text the same way the words I'm writing now are. I'd take a look at any community packages that you have installed that might be trying to run your code in real time or anything else, as those may be the culprits in crashing the editor as you type.

**Comment by ghost** [comment]
Well, I'm writing a new package from scratch.

There's an example here. On my system, if I install this package Pulsar crashes at startup. If you comment out the `import()` then it doesn't crash.

https://github.com/bojavou/pulsar-import

**Reply by confused-Techie** [reply]
~~In that case, from the sounds of the original message I was under the impression writing `import` crashes Pulsar. But knowing that you are creating a package that is using `import` I'd recommend reading up on the difference between ES Modules and CommonJS Modules.~~

~~Most NodeJS applications, including Pulsar use CommonJS, which generally, (as in by default) is mutually exclusive with ES Modules, which using `import` is a part of the ES Module System, rather than CommonJS. That's why in the code examples of creating a package it uses `require()` or within the codebase of Pulsar we use `require` rather than `import`.~~

~~There are some possibilities to look at to try and get this working, but nothing that I'm aware of that's mainlined for this purpose.~~

Sorry may have skimmed the message previously.

You do point out that you are using the `import()` function, which is in fact a way to utilize ES Modules within a CommonJS environment, sorry not super familiar on that function. Let me double check but my assumption, is that this is a feature within newer NodeJS versions than Pulsar is using. Because despite our best efforts Pulsar still uses a quite old NodeJS version, but it is something we are actively working to update. So let me double check that

**Reply by confused-Techie** [reply]
Alright so fully reading the docs, you are totally correct that according to the NodeJS v20.2.0 [docs](https://nodejs.org/api/esm.html#import-expressions) Dynamic imports are supporting within CommonJS when using `import()`. But looking back even the docs for NodeJS v16.20.0 say this should be possible, which is the version we are currently on. Do you have any logs from when it crashes? If you launch Pulsar via the CLI that will be able to show you all logs that occur until it crashes, to see exactly what's gone wrong.

**Reply by confused-Techie** [reply]
Additionally, so that I can test what's going wrong, could you provide a snippet of the code where this occurs?

**Reply by ghost** [reply]
There it is! Didn't even think to look at the output. The first run without the package succeeds and I closed it normally, then the second one with the package crashed.

```
[user@test-pulsar sft]$ ls
Pulsar-1.105.2023052200.AppImage
[user@test-pulsar sft]$ ./Pulsar-1.105.2023052200.AppImage --no-sandbox
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
[user@test-pulsar sft]$ ./Pulsar-1.105.2023052200.AppImage -p install https://github.com/bojavou/pulsar-import
Cloning https://github.com/bojavou/pulsar-import.git ✓
Installing modules ✓
Moving pulsar-import to /home/user/.pulsar/packages/pulsar-import ✓
[user@test-pulsar sft]$ ./Pulsar-1.105.2023052200.AppImage --no-sandbox
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
(node:2060) Electron: Loading non-context-aware native module in renderer: '/tmp/.org.chromium.Chromium.QJkBdI'. This is deprecated, see https://github.com/electron/electron/issues/18397.
(Use `exe --trace-warnings ...` to show where the warning was created)
(node:2060) Electron: Loading non-context-aware native module in renderer: '/tmp/.org.chromium.Chromium.SYUe6R'. This is deprecated, see https://github.com/electron/electron/issues/18397.
(node:2060) Electron: Loading non-context-aware native module in renderer: '/tmp/.org.chromium.Chromium.eLuYmI'. This is deprecated, see https://github.com/electron/electron/issues/18397.
(node:2060) Electron: Loading non-context-aware native module in renderer: '/tmp/.org.chromium.Chromium.l4fYIM'. This is deprecated, see https://github.com/electron/electron/issues/18397.
bash: cannot set terminal process group (-1): Inappropriate ioctl for device
bash: no job control in this shell
[user@test-pulsar sft]$ 
```

> Additionally, so that I can test what's going wrong, could you provide a snippet of the code where this occurs?

You could install the package, but if there's something I can do to make it easier I'm happy to whip it up.

This is the bit where it happens. I put it in `activate()` so it happens at startup time.

```js
module.exports.activate = async function () {
  const { value } = await import('./module.mjs')
}
```

**Reply by Daeraxa** [reply]
> But looking back even the docs for NodeJS v16.20.0 say this should be possible, 

I know we recommend to build with 16 but I've a suspicion that pulsar itself is using 14? Can't remember but also can't check atm

**Reply by mauricioszabo** [reply]
Pulsar is currently at node 14 indeed

**Comment by ghost** [comment]
Tried it with an internal module, to make sure it wasn't something about the relative path. This gives the same crash.

```
module.exports.activate = async function () {
  await import('querystring')
}
```
